lib/order.rb

```diff
@@ -0,0 +1,102 @@
+require 'csv'
+
+# Create a class called Order. Each new Order should include the following attributes:
+#   ID, a number (read-only)
+#   A collection of products and their cost
+#   An instance of Customer
+#   A fulfillment_status
+class Order
+  attr_reader :id
+  attr_accessor :products, :customer, :fulfillment_status
+  
+  def initialize (id, products, customer, fulfillment_status = :pending)  # If no fulfillment_status is provided, it will default to :pending
+    @id = id
+    @products = products
+    @customer = customer
+    @fulfillment_status = fulfillment_status
+    
+    # fulfillment_status, one of the following symbols :pending, :paid, :processing, :shipped, or :complete
+    valid_statuses = %i[pending paid processing shipped complete]
+    valid_statuses.include?(fulfillment_status)
+    
+    # If a status is given that is not one of the above, an ArgumentError should be raised
+    if !valid_statuses.include?(fulfillment_status)
+      raise ArgumentError.new "Invalid fulfillment status."
+    end
+  end
+  
+  
+  # Create a total method which will calculate the total cost of the order
+  def total
+    total_cost = 0
+    
+    # Summing up the product cost
+    @products.each do |item, cost|
+      total_cost += cost
+    end
+    
+    # Calculate a 7.5% tax
+    total_order_cost = total_cost * 1.075
+    
+    # Rounding the result to two decimal places
+    return total_order_cost.round(2)
+  end
+  
+  # Create an add_product method which will take in two parameters, product name and price, and add the data to the product collection
+  def add_product(product_name, price)
+    if @products.key?(product_name)
+      raise ArgumentError.new "Product of the same name had already been added to this order."
+    end
+    
+    @products[product_name] = price
+  end
+  
+  # Wave 2
+  
+  # Return a collection of Order instances, representing all of the Orders described in the CSV file
+  def self.all
+    orders = []
+    data = CSV.open("data/orders.csv").map(&:to_a).each do |row|
+
+      # @products contain product names and prices in the same column. Separate product name and price string into a hash
+      # Create array to hold set of product name and price after splitting by semicolon sign
+      product_array = []
+      product_split = row[1].split(';')
+
+      # Loop inside the array to get the string and split the product name and price by the colon sign
+      product_split.each do |product_string|
+        product_array << product_string.split(':')
+      end
+      
+      # Create hash of arrays. Each array is the product name and price pair
+      product_hash = Hash[product_array]
```

Comment:
I think this might be a little clearer since the `Hash[]` is kind of obscure:

```suggestion
      product_hash = product_array.to_h
```